File: ports/python/confine_core/certificate.py
"""Certificate issuance and verification (spec PROTOCOL.md sec 10).

Direct port of packages/confine/certificate.fard, transliterated from the
vector-confirmed Rust port. Uses the v1-fard MAC construction (tag
confine.certificate.mac.v1, key = raw UTF-8 bytes of the broker_secret
string) matching current real fardrun behavior.
"""

from dataclasses import dataclass, field
from .hash import certificate_mac_v1_fard
from .state import hash_state
from .capabilities import digest as capability_digest
from .policy import digest as policy_digest


def _label_to_dict(label):
    return {"kind": label.kind, "owner": label.owner, "compartments": list(label.compartments)}


def _action_to_dict(action: dict) -> dict:
    """Actions are plain dicts with FARD field names already, except any
    Label dataclass values (body_label, recipient_label, channel_label)
    need conversion before canonical encoding."""
    out = {}
    for k, v in action.items():
        out[k] = _label_to_dict(v) if hasattr(v, "kind") and hasattr(v, "owner") else v
    return out


def action_hash(action: dict) -> str:
    from .hash import tagged_digest
    return tagged_digest("confine.action.v1", _action_to_dict(action))


def _obligation_to_dict(o: dict) -> dict:
    out = dict(o)
    if "label" in out and hasattr(out["label"], "kind"):
        out["label"] = _label_to_dict(out["label"])
    return out


@dataclass
class Certificate:
    prior_state_hash: str
    action_hash: str
    actor_id: str
    policy_hash: str
    capability_hash: str
    nonce: str
    sequence: int
    obligations: list
    mac: str = ""


def _unsigned_dict(cert: Certificate) -> dict:
    return {
        "t": "transition_certificate",
        "version": 1,
        "prior_state_hash": cert.prior_state_hash,
        "action_hash": cert.action_hash,
        "actor_id": cert.actor_id,
        "policy_hash": cert.policy_hash,
        "capability_hash": cert.capability_hash,
        "nonce": cert.nonce,
        "sequence": cert.sequence,
        "obligations": [_obligation_to_dict(o) for o in cert.obligations],
    }


def issue(state, action, actor_id, capabilities, policy, obligations, broker_secret) -> Certificate:
    unsigned = Certificate(
        prior_state_hash=hash_state(state),
        action_hash=action_hash(action),
        actor_id=actor_id,
        policy_hash=policy_digest(policy),
        capability_hash=capability_digest(capabilities),
        nonce=action["nonce"],
        sequence=state.sequence,
        obligations=obligations,
    )
    mac = certificate_mac_v1_fard(broker_secret, _unsigned_dict(unsigned))
    unsigned.mac = mac
    return unsigned


class VerifyError(Exception):
    def __init__(self, code):
        self.code = code
        super().__init__(code)


def verify(certificate: Certificate, state, action, actor_id, capabilities, policy, broker_secret):
    if not certificate.mac:
        return VerifyError("CERTIFICATE_MISSING_MAC")
    expected_mac = certificate_mac_v1_fard(broker_secret, _unsigned_dict(certificate))
    if certificate.mac != expected_mac:
        return VerifyError("CERTIFICATE_MAC_MISMATCH")
    if certificate.prior_state_hash != hash_state(state):
        return VerifyError("CERTIFICATE_STATE_MISMATCH")
    if certificate.action_hash != action_hash(action):
        return VerifyError("CERTIFICATE_ACTION_MISMATCH")
    if certificate.actor_id != actor_id:
        return VerifyError("CERTIFICATE_ACTOR_MISMATCH")
    if certificate.policy_hash != policy_digest(policy):
        return VerifyError("CERTIFICATE_POLICY_MISMATCH")
    if certificate.capability_hash != capability_digest(capabilities):
        return VerifyError("CERTIFICATE_CAPABILITY_MISMATCH")
    if certificate.nonce != action["nonce"]:
        return VerifyError("CERTIFICATE_NONCE_MISMATCH")
    if certificate.sequence != state.sequence:
        return VerifyError("CERTIFICATE_SEQUENCE_MISMATCH")
    return None

File: ports/js/src/certificate.js
// Certificate issuance and verification (spec PROTOCOL.md sec 10).
//
// Direct port of packages/confine/certificate.fard, transliterated from
// the vector-confirmed Rust and Python ports. Uses the v1-fard MAC
// construction (tag confine.certificate.mac.v1, key = raw UTF-8 bytes of
// the broker_secret string) matching current real fardrun behavior.

import { certificateMacV1Fard, taggedDigest } from "./hash.js";
import { hashState } from "./state.js";
import { capabilitiesDigest } from "./capabilities.js";
import { policyDigest } from "./policy.js";

function isLabel(v) {
  return v && typeof v === "object" && "kind" in v && "owner" in v;
}

function actionToValue(action) {
  const out = {};
  for (const [k, v] of Object.entries(action)) {
    out[k] = isLabel(v) ? { kind: v.kind, owner: v.owner, compartments: [...v.compartments] } : v;
  }
  return out;
}

export function actionHash(action) {
  return taggedDigest("confine.action.v1", actionToValue(action));
}

function obligationToValue(o) {
  const out = { ...o };
  if (isLabel(out.label)) {
    out.label = { kind: out.label.kind, owner: out.label.owner, compartments: [...out.label.compartments] };
  }
  return out;
}

function unsignedValue(cert) {
  return {
    t: "transition_certificate",
    version: 1,
    prior_state_hash: cert.priorStateHash,
    action_hash: cert.actionHash,
    actor_id: cert.actorId,
    policy_hash: cert.policyHash,
    capability_hash: cert.capabilityHash,
    nonce: cert.nonce,
    sequence: cert.sequence,
    obligations: cert.obligations.map(obligationToValue),
  };
}

export function issue(state, action, actorId, capabilities, policy, obligations, brokerSecret) {
  const unsigned = {
    priorStateHash: hashState(state),
    actionHash: actionHash(action),
    actorId,
    policyHash: policyDigest(policy),
    capabilityHash: capabilitiesDigest(capabilities),
    nonce: action.nonce,
    sequence: state.sequence,
    obligations,
  };
  const mac = certificateMacV1Fard(brokerSecret, unsignedValue(unsigned));
  return { ...unsigned, mac };
}

export function verify(certificate, state, action, actorId, capabilities, policy, brokerSecret) {
  if (!certificate.mac) return "CERTIFICATE_MISSING_MAC";
  const expectedMac = certificateMacV1Fard(brokerSecret, unsignedValue(certificate));
  if (certificate.mac !== expectedMac) return "CERTIFICATE_MAC_MISMATCH";
  if (certificate.priorStateHash !== hashState(state)) return "CERTIFICATE_STATE_MISMATCH";
  if (certificate.actionHash !== actionHash(action)) return "CERTIFICATE_ACTION_MISMATCH";
  if (certificate.actorId !== actorId) return "CERTIFICATE_ACTOR_MISMATCH";
  if (certificate.policyHash !== policyDigest(policy)) return "CERTIFICATE_POLICY_MISMATCH";
  if (certificate.capabilityHash !== capabilitiesDigest(capabilities)) return "CERTIFICATE_CAPABILITY_MISMATCH";
  if (certificate.nonce !== action.nonce) return "CERTIFICATE_NONCE_MISMATCH";
  if (certificate.sequence !== state.sequence) return "CERTIFICATE_SEQUENCE_MISMATCH";
  return null;
}

export { obligationToValue };
